File: 프로그래머스/lv2/87946. 피로도/피로도.java
// 순열 구합니다.
// 순열을 순회합니다.
//      최대 던전 수를 구합니다.
//      max = Math.max(max, "최대 던전 수");
// max를 반환합니다.

import java.util.*;

class Solution {
    private void permutations(ArrayList<Integer> list, int k, int[][] dungeons, int size, int depth, int count, boolean[] visited) {
        if (size == depth) {
            list.add(count);
            return;
        }
        
        for (int i = 0; i < dungeons.length; i++) {
            if (visited[i]) continue;
            if (dungeons[i][0] > k) {
                list.add(count);
                continue;
            }
            visited[i] = true;
            permutations(list, k - dungeons[i][1], dungeons, size, depth + 1, count + 1, visited);
            visited[i] = false;
        }
    }
    
    public int solution(int k, int[][] dungeons) {
        ArrayList<Integer> countList = new ArrayList<>(){{
            add(0);
        }};
        boolean[] visited = new boolean[dungeons.length];
        permutations(countList, k, dungeons, dungeons.length, 0, 0, visited);
        
        return Collections.max(countList);
    }
}

File: 프로그래머스/lv2/87946. 피로도/피로도.js
function getPermitations(arr) {
    function swap(i, j) {
        const temp = arr[i]
        arr[i] = arr[j]
        arr[j] = temp
    }
    
    function makePermutations(start) {
        if (start === arr.length) {
            result.push([...arr])
            return
        }

        for (let i = start; i < arr.length; i++) {
            swap(start, i);
            makePermutations(start + 1)
            swap(start, i);
        }        
    }
    
    result = []
    makePermutations(0)
    return result
}


function solution(k, dungeons) {
    function getCanEnteredDungeonsCount(d) {
        let hp = k
        let count = 0
        for (const [need, pay] of d) {
            if (hp < need) break
            hp -= pay
            count += 1
        }
    
        return count
    }
    
    let answer = 0;
    for (const dungeon of getPermitations(dungeons)) {
        answer = Math.max(answer, getCanEnteredDungeonsCount(dungeon))
    }
    return answer;
}